services/worker/src/worker/executor.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 The HuggingFace Authors.
+import asyncio
+import json
+import logging
+import os
+import sys
+from typing import Any, Callable, Optional
+
+from filelock import FileLock
+from libcommon.queue import Queue
+from mirakuru import OutputExecutor
+
+from worker import start_worker_loop
+from worker.config import AppConfig
+from worker.job_runner_factory import JobRunnerFactory
+from worker.loop import WorkerState
+
+START_WORKER_LOOP_PATH = start_worker_loop.__file__
+
+
+async def every(
+    func: Callable[..., Optional[Any]], *args: Any, seconds: int, stop_on: Optional[Any] = None, **kwargs: Any
+) -> None:
+    while True:
+        out = func(*args, **kwargs)
+        if stop_on is not None and out == stop_on:
+            break
+        await asyncio.sleep(seconds)
+
+
+class WorkerExecutor:
+    def __init__(self, app_config: AppConfig, job_runner_factory: JobRunnerFactory) -> None:
+        self.app_config = app_config
+        self.job_runner_factory = job_runner_factory
+
+    def _create_worker_loop_executor(self) -> OutputExecutor:
+        banner = self.app_config.worker.state_path
+        if not banner:
+            raise ValueError("Failed to create the executor because WORKER_STATE_PATH is missing.")
+        start_worker_loop_command = [
+            sys.executable,
+            START_WORKER_LOOP_PATH,
+            "--print-worker-state-path",
+        ]
+        return OutputExecutor(start_worker_loop_command, banner, timeout=10)
+
+    def start(self) -> None:
+        worker_loop_executor = self._create_worker_loop_executor()
+        worker_loop_executor.start()  # blocking until the banner is printed
+
+        loop = asyncio.get_event_loop()
+        logging.info("Starting heartbeat.")
+        loop.create_task(every(self.heartbeat, seconds=self.app_config.worker.heartbeat_interval_seconds))
+        loop.create_task(every(self.kill_zombies, seconds=self.app_config.worker.kill_zombies_interval_seconds))
```

Comment:
hmmm: it means that all the workers (there can be hundreds of them) are doing the same job of killing zombies. 

workers vs zombies

I'm not sure it's an issue though (maybe race condition?)

Reply:
Yea it will spam the database a bit so I'd be in favor of mo

Reply:
Yes, I think it's ok for now. We can let a comment in the code, or create an issue, to move it one day to a "maintenance" job

Reply:
or not